oauth2_http/java/com/google/auth/oauth2/ComputeEngineCredentials.java

```diff
@@ -192,6 +201,61 @@ String createTokenUrlWithScopes() {
     return tokenUrl.toString();
   }
 
+  /**
+   * Gets the universe domain from the GCE metadata server.
+   *
+   * <p>Returns an explicit universe domain if it was provided during credential initialization.
+   *
+   * <p>Returns the {@link Credentials#GOOGLE_DEFAULT_UNIVERSE} if universe domain endpoint is
+   * unavailable or returns an empty string.
+   *
+   * <p>Otherwise, returns universe domain from GCE metadata service.
+   *
+   * <p>Any above value is cached for the credential lifetime.
+   *
+   * @throws IOException if a call to GCE metadata service was unsuccessful. Check if exception
+   *     implements the {@link Retryable} and {@code isRetryable()} will return true if the
+   *     operation may be retried.
+   * @return string representing a universe domain in the format some-domain.xyz
+   */
+  @Override
+  public String getUniverseDomain() throws IOException {
+    if (!isDefaultUniverseDomain()) {
+      return super.getUniverseDomain();
+    }
+
+    synchronized (this) {
+      if (universeDomainFromMetadata != null) {
+        return universeDomainFromMetadata;
+      }
+
+      universeDomainFromMetadata = getUniverseDomainFromMetadata();
```

Comment:
We should avoid holding the mutex while making a network call.  Since the response should always provide the same universe, we could:

```
synch(this) {
  if (this.uDFM != null) { 
    return this.uDFM;
  }
}

String uDFM = getUDFM();

synchronized (this) {
  this.uDFM = uDFM;
}
return uDFM;

Reply:
thanks, makes sense )